### ISSUE TITLE
Possible missing causal masking for WKV kernel

### ISSUE THREAD

**Issue by Smith42** [Hannibal046/nanoRWKV#3]
Hi, I have been testing this code on a set of GPUs, and am getting some results that seem a little *too* good! Much faster training when compared to an equivilent GPT-2 model (please see attached loss plot).

On inspecting the code more closely it seems like the WKV kernel is missing the causal masking that is used in decoding transformers to stop the model from looking ahead in a sequence. I am new to the RWKV arch, so am not sure how to fix this, but could you take a look and see if my hunch is correct?

[loss_compare_RWKV.pdf](https://github.com/Hannibal046/nanoRWKV/files/11923122/loss_compare_RWKV.pdf)

**Comment by Hannibal046** [comment]
Hi, thanks for testing this code. I would explain point by point:
- good loss curve: actually, i think LR and optimizer have huge impact on the loss curve beside the model architecture. In the current implementation, following [official implementation](https://github.com/BlinkDL/RWKV-LM/blob/main/RWKV-v4neo/src/model.py) i didn't introduce dropout which may lead to overfitting to training set. For the optimization, the current code uses the same as nanoGPT. 
- faster training: this is not expected, on my side the training of rwkv is a little bit slower than that of GPT-2. I am using 8*v100 32G. May i know your computing device?
- casual masking: the current code for training and single batch decoding doesn't need casual mask since WKV kernel is essentially an RNN. You could refer to this [huggingface implementation](https://github.com/huggingface/transformers/blob/d51aa48a76b470d95fe665bbe7ee97a3d4c21fb9/src/transformers/models/rwkv/modeling_rwkv.py#L565-L571) which also doesn't use attention mask for now.

**Comment by Smith42** [comment]
Hi, thanks for the quick reply.

* I think the loss curves look to be levelling out at around the same loss as my GPT runs, which I suppose is reassuring (if a little disappointing!).
* The training time per batch is around the same as my nanoGPT runs, but the loss decrease on RWKV was far faster per data item which is what surprised me. I updated the CUDA stack on my machine in between the GPT and RWKV runs which may have changed things, so will replicate my GPT 700M run when RWKV is done and see if they match up again.

I'll close this issue as it looks like this was me misunderstanding the RWKV arch. I will go read up on the theory for a bit longer!

To answer your question; I'm running this on 8x40G A100 machines. I can also confirm that bfloat16 is working well for me, so that can be ticked off your TODO list.
